--- a/slam_gmapping/include/slam_gmapping/slam_gmapping.h
+++ b/slam_gmapping/include/slam_gmapping/slam_gmapping.h
@@ -90,6 +90,7 @@
     nav_msgs::msg::OccupancyGrid map_;
 
     tf2::Duration map_update_interval_;
+    tf2::TimePoint last_map_update_;
     tf2::Transform map_to_odom_;
     std::mutex map_to_odom_mutex_;
     std::mutex map_mutex_;
--- a/slam_gmapping/src/slam_gmapping.cpp
+++ b/slam_gmapping/src/slam_gmapping.cpp
@@ -62,7 +62,10 @@
     map_frame_ = "map";
     odom_frame_ = "odom";
-    transform_publish_period_ = 0.05;
+    transform_publish_period_ =
+        this->declare_parameter<double>("transform_publish_period", 0.05);
 
-    map_update_interval_ = tf2::durationFromSec(0.5);
+    map_update_interval_ = tf2::durationFromSec(
+        this->declare_parameter<double>("map_update_interval", 0.5));
+    last_map_update_ = tf2::TimePointZero;
     maxUrange_ = 80.0;  maxRange_ = 0.0;
     minimum_score_ = 0;
@@ -354,8 +357,6 @@
     if ((laser_count_ % throttle_scans_) != 0)
         return;
 
-    tf2::TimePoint last_map_update = tf2::TimePointZero;
-
     // We can't initialize the mapper until we've got the first scan
     if(!got_first_scan_)
     {
@@ -381,10 +382,10 @@
         map_to_odom_mutex_.unlock();
 
         tf2::TimePoint timestamp = tf2_ros::fromMsg(scan->header.stamp);
-        if(!got_map_ || (timestamp - last_map_update) > map_update_interval_)
+        if(!got_map_ || (timestamp - last_map_update_) > map_update_interval_)
         {
             updateMap(scan);
-            last_map_update = tf2_ros::fromMsg(scan->header.stamp);
+            last_map_update_ = timestamp;
         }
     }
 }
@@ -513,8 +514,7 @@
 void SlamGmapping::publishTransform()
 {
     map_to_odom_mutex_.lock();
-    rclcpp::Time tf_expiration = get_clock()->now() + rclcpp::Duration(
-            static_cast<int32_t>(static_cast<rcl_duration_value_t>(tf_delay_)), 0);
+    rclcpp::Time tf_expiration = get_clock()->now() + rclcpp::Duration::from_seconds(tf_delay_);
     geometry_msgs::msg::TransformStamped transform;
     transform.header.frame_id = map_frame_;
     transform.header.stamp = tf_expiration;
